File: cpp/tests/test_config.cpp
#include <doctest/doctest.h>

#include <filesystem>
#include <fstream>
#include <string>
#include <vector>

#include "core/config.hpp"
#include "core/errors.hpp"
#include "util/hex.hpp"

using namespace erikslund;

TEST_CASE("defaults are the regtest harness values") {
    const Config config = Config::from_string("{}");
    CHECK(config.rpc_url == "http://127.0.0.1:18443");
    CHECK(config.bind_port == 3333);
    CHECK(config.extranonce2_size == 8);
    CHECK(config.coinbase_version == 1);
}

TEST_CASE("known keys override, others keep defaults") {
    const Config config = Config::from_string(R"({
        "bitcoin_nodes": [{"address": "http://node:8332", "username": "u", "password": "p"}],
        "stratum_listen": ["0.0.0.0:4444"],
        "initial_difficulty": 1024,
        "coinbase_signature": "/test/"
    })");
    CHECK(config.rpc_url == "http://node:8332");
    CHECK(config.rpc_user == "u");
    CHECK(config.bind_port == 4444);
    CHECK(config.initial_difficulty == doctest::Approx(1024.0));
    CHECK(config.coinbase_signature == "/test/");
    CHECK(config.api_port == 7777); // untouched default
}

TEST_CASE("the pool config schema parses into the typed config") {
    const Config c = Config::from_string(R"({
        "bitcoin_nodes": [
            {"address": "bitcoind:8332", "username": "u", "password": "p", "notify": true},
            {"address": "backup:8332", "username": "u2", "password": "p2"}
        ],
        "stratum_listen": ["0.0.0.0:3333", "0.0.0.0:4001"],
        "coinbase_signature": "/erikslund/",
        "initial_difficulty": 10000,
        "minimum_difficulty": 1,
        "maximum_difficulty": 1000000,
        "extranonce1_size": 4,
        "extranonce2_size": 8,
        "api_listen": ["127.0.0.1:7777"],
        "zmq_block_endpoint": "tcp://bitcoind:28332",
        "block_poll_milliseconds": 250,
        "work_rebroadcast_seconds": 20,
        "max_clients": 50,
        "drop_idle_seconds": 900,
        "auth_timeout_seconds": 45,
        "max_protocol_errors": 7,
        "worker_threads": 6,
        "version_rolling_mask": "1fffe000"
    })");
    CHECK(c.rpc_url == "bitcoind:8332"); // scheme added later by RpcClient
    CHECK(c.rpc_user == "u");
    CHECK(c.rpc_password == "p");
    REQUIRE(c.rpc_failover.size() == 1);
    CHECK(c.rpc_failover[0].url == "backup:8332");
    CHECK(c.rpc_endpoints().size() == 2);
    CHECK(c.stratum_ports() == std::vector<uint16_t>{3333, 4001});
    CHECK(c.bind_host == "0.0.0.0");
    CHECK(c.coinbase_signature == "/erikslund/");
    CHECK(c.initial_difficulty == doctest::Approx(10000.0));
    CHECK(c.minimum_difficulty == doctest::Approx(1.0));
    CHECK(c.maximum_difficulty == doctest::Approx(1000000.0));
    CHECK(c.extranonce1_size == 4);
    CHECK(c.extranonce2_size == 8);
    CHECK(c.api_host == "127.0.0.1");
    CHECK(c.api_port == 7777);
    CHECK(c.zmq_block_endpoint == "tcp://bitcoind:28332");
    CHECK(c.poll_interval == doctest::Approx(0.25)); // block_poll_milliseconds -> seconds
    CHECK(c.work_rebroadcast_seconds == doctest::Approx(20.0));
    CHECK(c.max_clients == 50);
    CHECK(c.drop_idle_seconds == 900);
    CHECK(c.auth_timeout_seconds == 45);
    CHECK(c.max_protocol_errors == 7);
    CHECK(c.worker_threads == 6);
    CHECK(c.version_rolling_mask == 0x1fffe000u);
}

TEST_CASE("a single stratum_listen string is accepted") {
    const Config config = Config::from_string(R"({"stratum_listen": "0.0.0.0:3333"})");
    CHECK(config.bind_host == "0.0.0.0");
    CHECK(config.stratum_ports() == std::vector<uint16_t>{3333});
}

TEST_CASE("SV2 plaintext listeners are disabled by default and default to loopback") {
    CHECK(Config::from_string("{}").sv2_plaintext_ports.empty());

    const Config config = Config::from_string(
        R"({"sv2_plaintext_listen": [":3334", ":3335"]})");
    CHECK(config.sv2_plaintext_host == "127.0.0.1");
    CHECK(config.sv2_plaintext_ports == std::vector<uint16_t>{3334, 3335});
}

TEST_CASE("SV2 plaintext listeners reject mixed hosts") {
    CHECK_THROWS_AS(
        Config::from_string(
            R"({"sv2_plaintext_listen": ["127.0.0.1:3334", "0.0.0.0:3335"]})"),
        ConfigError);
    CHECK_THROWS_AS(
        Config::from_string(R"({"sv2_plaintext_listen": "0.0.0.0:3334"})"),
        ConfigError);
}

TEST_CASE("SV2 Noise listeners require a complete responder credential set") {
    CHECK(Config::from_string("{}").sv2_ports.empty());
    CHECK_THROWS_AS(
        Config::from_string(R"({"sv2_listen": ":3334"})"),
        ConfigError);
    CHECK_THROWS_AS(
        Config::from_string(
            R"({"sv2_static_secret_key_file": "static.key"})"),
        ConfigError);

    const Config config = Config::from_string(R"({
        "sv2_listen": [":3334", ":3335"],
        "sv2_static_secret_key_file": "/run/secrets/sv2-static",
        "sv2_authority_public_key_file": "/run/secrets/sv2-authority",
        "sv2_certificate_file": "/run/secrets/sv2-certificate"
    })");
    CHECK(config.sv2_host == "0.0.0.0");
    CHECK(config.sv2_ports == std::vector<uint16_t>{3334, 3335});
    CHECK(config.sv2_static_secret_key_file == "/run/secrets/sv2-static");
}

TEST_CASE("SV2 Noise listeners reject mixed hosts") {
    CHECK_THROWS_AS(
        Config::from_string(R"({
            "sv2_listen": ["127.0.0.1:3334", "0.0.0.0:3335"],
            "sv2_static_secret_key_file": "static.key",
            "sv2_authority_public_key_file": "authority.pub",
            "sv2_certificate_file": "certificate.bin"
        })"),
        ConfigError);
}

TEST_CASE("authenticated and plaintext SV2 listeners must use distinct ports") {
    const std::string credentials = R"(
        "sv2_static_secret_key_file": "static.key",
        "sv2_authority_public_key_file": "authority.pub",
        "sv2_certificate_file": "certificate.bin"
    )";

    CHECK_THROWS_AS(
        Config::from_string(
            R"({"sv2_listen": "127.0.0.1:3334",
                "sv2_plaintext_listen": "127.0.0.1:3334",)" +
            credentials + "}"),
        ConfigError);
    CHECK_THROWS_AS(
        Config::from_string(
            R"({"sv2_listen": ":3334",
                "sv2_plaintext_listen": ":3334",)" +
            credentials + "}"),
        ConfigError);

    CHECK_THROWS_AS(
        Config::from_string(
            R"({"sv2_listen": "192.0.2.1:3334",
                "sv2_plaintext_listen": "127.0.0.1:3334",)" +
            credentials + "}"),
        ConfigError);
    CHECK_NOTHROW(Config::from_string(
        R"({"sv2_listen": "127.0.0.1:3334",
            "sv2_plaintext_listen": "127.0.0.1:13334",)" +
        credentials + "}"));
}

TEST_CASE("proxy_protocol_from: absent -> empty (feature off)") {
    CHECK(Config::from_string("{}").proxy_protocol_from.empty());
}

TEST_CASE("proxy_protocol_from accepts a single string") {
    const Config c = Config::from_string(R"({"proxy_protocol_from": "10.0.0.5"})");
    REQUIRE(c.proxy_protocol_from.size() == 1);
    CHECK(c.proxy_protocol_from[0] == "10.0.0.5");
}

TEST_CASE("proxy_protocol_from accepts a list, in order") {
    const Config c = Config::from_string(R"({"proxy_protocol_from": ["10.0.0.5", "172.19.0.0/16"]})");
    REQUIRE(c.proxy_protocol_from.size() == 2);
    CHECK(c.proxy_protocol_from[0] == "10.0.0.5");
    CHECK(c.proxy_protocol_from[1] == "172.19.0.0/16");
}

TEST_CASE("the old flat schema is rejected (no backwards-compat bridge)") {
    CHECK_THROWS_AS(Config::from_string(R"({"rpc_url": "http://node:8332", "bind_port": 4444})"),
                    ConfigError);
}

TEST_CASE("unknown keys are rejected") {
    CHECK_THROWS_AS(Config::from_string(R"({"definitely_not_a_key": 1})"), ConfigError);
}

TEST_CASE("a non-object is rejected") {
    CHECK_THROWS_AS(Config::from_string("[]"), ConfigError);
}

TEST_CASE("donation_percent must be within [0, 100]") {
    CHECK_THROWS_AS(Config::from_string(R"({"donation_percent": -1.0})"), ConfigError);
    CHECK_THROWS_AS(Config::from_string(R"({"donation_percent": 100.5})"), ConfigError);
    // Boundaries are accepted (a non-zero percent needs a donation_address; 0 does not).
    CHECK(Config::from_string(R"({"donation_percent": 0.0})").donation_percent == 0.0);
    CHECK(Config::from_string(R"({"donation_percent": 100.0, "donation_address": "bc1qexample"})")
              .donation_percent == doctest::Approx(100.0));
    CHECK(Config::from_string(R"({"donation_percent": 1.5, "donation_address": "bc1qexample"})")
              .donation_percent == doctest::Approx(1.5));
}

TEST_CASE("an over-long coinbase_signature is rejected at load") {
    // Budget 10 + 4 + 8 + tag <= 100, so the default extranonces leave the tag 78 bytes.
    CHECK_THROWS_AS(
        Config::from_string(std::string(R"({"coinbase_signature": ")") + std::string(79, 'x') +
                            R"("})"),
        ConfigError);
    CHECK(Config::from_string(std::string(R"({"coinbase_signature": ")") + std::string(78, 'x') +
                              R"("})")
              .coinbase_signature.size() == 78);
}

TEST_CASE("work_source selects IPC without a separate acknowledgement") {
    CHECK(Config::from_string(R"({"work_source": "ipc"})").work_source == "ipc");
    CHECK(Config::from_string(R"({"work_source": "rpc"})").work_source == "rpc");
    CHECK_THROWS_AS(Config::from_string(R"({"experimental_ipc": true})"), ConfigError);
}

namespace {
// Caller removes the returned path.
std::filesystem::path write_temp(const std::string& name, const std::string& text) {
    const std::filesystem::path path = std::filesystem::temp_directory_path() / name;
    std::ofstream(path, std::ios::binary) << text;
    return path;
}
} // namespace

TEST_CASE("from_file parses a YAML config") {
    // Numbers and booleans stay typed; strings that look numeric are quoted.
    const auto path = write_temp("ep_config_test.yaml", R"(
bitcoin_nodes:
  - address: http://node:8332
    username: u
    password: p
stratum_listen:
  - 0.0.0.0:4444
initial_difficulty: 1024
minimum_difficulty: 0.0001
block_poll_milliseconds: 250
variable_difficulty: false
coinbase_signature: "/test/"
version_rolling_mask: 1fffe000
)");
    const Config config = Config::from_file(path.string());
    CHECK(config.rpc_url == "http://node:8332");
    CHECK(config.rpc_user == "u");
    CHECK(config.bind_port == 4444);
    CHECK(config.initial_difficulty == doctest::Approx(1024.0));
    // A fractional scalar must stay a double, not be truncated to an int (0).
    CHECK(config.minimum_difficulty == doctest::Approx(0.0001));
    CHECK(config.poll_interval == doctest::Approx(0.25)); // 250 ms -> seconds
    CHECK(config.variable_difficulty == false);
    CHECK(config.coinbase_signature == "/test/");
    // A bare hex string (not numeric) stays a string and is parsed base-16.
    CHECK(config.version_rolling_mask == 0x1fffe000u);
    std::filesystem::remove(path);
}

TEST_CASE("max_workers_per_address loads, defaults to 256, and rejects negatives") {
    CHECK(Config::from_string(R"({"max_workers_per_address": 32})").max_workers_per_address == 32);
    CHECK(Config::from_string("{}").max_workers_per_address == 256); // default
    CHECK(Config::from_string(R"({"max_workers_per_address": 0})").max_workers_per_address == 0);
    CHECK_THROWS_AS(Config::from_string(R"({"max_workers_per_address": -1})"), ConfigError);
}

TEST_CASE("user_stats_retention_days loads, defaults to 90, and rejects negatives") {
    CHECK(Config::from_string(R"({"user_stats_retention_days": 30})").user_stats_retention_days == 30);
    CHECK(Config::from_string("{}").user_stats_retention_days == 90); // default
    CHECK(Config::from_string(R"({"user_stats_retention_days": 0})").user_stats_retention_days == 0);
    CHECK_THROWS_AS(Config::from_string(R"({"user_stats_retention_days": -1})"), ConfigError);
    // Upper bound (~100 years): beyond it the file_time_type cutoff arithmetic would overflow.
    CHECK_THROWS_AS(Config::from_string(R"({"user_stats_retention_days": 36501})"), ConfigError);
}

TEST_CASE("YAML quoting is honored: quoted scalars stay strings (matches the Python pool)") {
    // A quoted numeric-looking value is an explicit string. For a string-typed key it must
    // survive untouched (a password like "12345" or "true" must never coerce to a number/bool).
    const auto ok_path = write_temp("ep_config_quote_ok.yaml", R"(
coinbase_signature: "12345"
version_rolling_mask: "1fffe000"
)");
    const Config config = Config::from_file(ok_path.string());
    CHECK(config.coinbase_signature == "12345");
    CHECK(config.version_rolling_mask == 0x1fffe000u); // quoted mask string still parses base-16
    std::filesystem::remove(ok_path);

    // For a numeric-typed key a quoted value is a type error, not a silent coercion.
    const auto bad_path = write_temp("ep_config_quote_bad.yaml", R"(
initial_difficulty: "1024"
)");
    CHECK_THROWS_AS(Config::from_file(bad_path.string()), ConfigError);
    std::filesystem::remove(bad_path);
}

TEST_CASE("from_file still reads a JSON config (YAML is a JSON superset)") {
    const auto path = write_temp("ep_config_test.json", R"({
        "bitcoin_nodes": [{"address": "http://node:8332"}],
        "stratum_listen": "0.0.0.0:3333",
        "max_clients": 50
    })");
    const Config config = Config::from_file(path.string());
    CHECK(config.rpc_url == "http://node:8332");
    CHECK(config.bind_port == 3333);
    CHECK(config.max_clients == 50);
    std::filesystem::remove(path);
}

TEST_CASE("from_file throws ConfigError on malformed YAML") {
    const auto path = write_temp("ep_config_bad.yaml", "bitcoin_nodes: [unterminated\n");
    CHECK_THROWS_AS(Config::from_file(path.string()), ConfigError);
    std::filesystem::remove(path);
}

TEST_CASE("from_file throws ConfigError when the file cannot be opened") {
    CHECK_THROWS_AS(Config::from_file("/nonexistent/dir/definitely_absent.yaml"), ConfigError);
}

TEST_CASE("version_rolling_mask accepts both a hex string and a JSON integer") {
    // Bare hex string -> parsed base-16.
    CHECK(Config::from_string(R"({"version_rolling_mask": "1fffe000"})").version_rolling_mask ==
          0x1fffe000u);
    // JSON integer -> taken as-is (within the BIP320 range, no clamp). 0x2000 = 8192.
    CHECK(Config::from_string(R"({"version_rolling_mask": 8192})").version_rolling_mask ==
          0x00002000u);
}

TEST_CASE("extranonce1 prefix decodes from config and leaves a four-byte counter") {
    const Config config = Config::from_string(
        R"({"extranonce1_size": 6, "extranonce1_prefix": "A0b1"})");
    CHECK(util::to_hex(config.extranonce1_prefix) == "a0b1");
    CHECK(config.extranonce1_size - config.extranonce1_prefix.size() == 4);
}

TEST_CASE("version_rolling_mask is clamped to the BIP320 range") {
    // Bits outside 0x1fffe000 would let a miner roll nVersion's high bits -> bad-version blocks.
    CHECK(Config::from_string(R"({"version_rolling_mask": "ffffffff"})").version_rolling_mask ==
          0x1fffe000u);
    // 0x3000 = 12288; bit 12 dropped, bit 13 kept -> 0x2000.
    CHECK(Config::from_string(R"({"version_rolling_mask": 12288})").version_rolling_mask ==
          0x00002000u);
}

TEST_CASE("version_rolling_mask must retain a BIP320 bit") {
    CHECK_THROWS_AS(Config::from_string(R"({"version_rolling_mask": 0})"), ConfigError);
    CHECK_THROWS_AS(Config::from_string(R"({"version_rolling_mask": "00000000"})"),
                    ConfigError);
    CHECK_THROWS_AS(Config::from_string(R"({"version_rolling_mask": 4096})"), ConfigError);
}

TEST_CASE("every remaining scalar key maps onto its config field") {
    const Config c = Config::from_string(R"({
        "coinbase_version": 2,
        "variable_difficulty": false,
        "vardiff_target_shares_per_minute": 30,
        "vardiff_retarget_seconds": 120,
        "fast_block_notify": false,
        "status_interval_seconds": 15,
        "stats_directory": "/var/lib/pool",
        "max_line_bytes": 8192,
        "donation_percent": 2.5,
        "donation_address": "bc1qexample"
    })");
    CHECK(c.coinbase_version == 2u);
    CHECK(c.variable_difficulty == false);
    CHECK(c.vardiff_target_shares_per_minute == doctest::Approx(30.0));
    CHECK(c.vardiff_retarget_seconds == 120);
    CHECK(c.fast_block_notify == false);
    CHECK(c.status_interval_seconds == doctest::Approx(15.0));
    CHECK(c.stats_directory == "/var/lib/pool");
    CHECK(c.max_line_bytes == 8192u);
    CHECK(c.donation_percent == doctest::Approx(2.5));
    CHECK(c.donation_address == "bc1qexample");
}

TEST_CASE("api_listen accepts a single string and an array (first entry wins)") {
    CHECK(Config::from_string(R"({"api_listen": "127.0.0.1:9000"})").api_port == 9000);
    const Config c = Config::from_string(R"({"api_listen": ["0.0.0.0:9001", "ignored:1"]})");
    CHECK(c.api_host == "0.0.0.0");
    CHECK(c.api_port == 9001);
}

TEST_CASE("an empty bitcoin_nodes array leaves the RPC defaults in place") {
    const Config c = Config::from_string(R"({"bitcoin_nodes": []})");
    CHECK(c.rpc_url == "http://127.0.0.1:18443"); // default preserved
    CHECK(c.rpc_failover.empty());
}

TEST_CASE("a failover node inherits the primary's credentials when omitted") {
    const Config c = Config::from_string(R"({
        "bitcoin_nodes": [
            {"address": "http://primary:8332", "username": "alice", "password": "secret"},
            {"address": "http://backup:8332"}
        ]
    })");
    REQUIRE(c.rpc_failover.size() == 1);
    CHECK(c.rpc_failover[0].user == "alice");
    CHECK(c.rpc_failover[0].password == "secret");
}

TEST_CASE("block_poll_milliseconds converts to a fractional second poll interval") {
    CHECK(Config::from_string(R"({"block_poll_milliseconds": 500})").poll_interval ==
          doctest::Approx(0.5));
    CHECK(Config::from_string(R"({"block_poll_milliseconds": 1000})").poll_interval ==
          doctest::Approx(1.0));
}

TEST_CASE("worker_threads 0 (auto) is a valid value that survives parsing") {
    CHECK(Config::from_string(R"({"worker_threads": 0})").worker_threads == 0);
    CHECK(Config::from_string(R"({"worker_threads": 12})").worker_threads == 12);
}

TEST_CASE("max_line_bytes must be positive") {
    CHECK_THROWS_AS(Config::from_string(R"({"max_line_bytes": 0})"), ConfigError);
    CHECK_THROWS_AS(
        Config::from_string(R"({"max_line_bytes": 16777216})"),
        ConfigError);
}

TEST_CASE("the scriptSig budget tracks the extranonce sizes, not just the signature length") {
    // height_push(10) + en1 + en2 + sig must stay <= 100. With en1=8, en2=8, a 75-byte
    // tag = 10+8+8+75 = 101 -> rejected; 74 -> 100 exactly -> accepted.
    CHECK_THROWS_AS(
        Config::from_string(std::string(R"({"extranonce1_size": 8, "extranonce2_size": 8, )"
                                        R"("coinbase_signature": ")") +
                            std::string(75, 'x') + R"("})"),
        ConfigError);
    CHECK_NOTHROW(
        Config::from_string(std::string(R"({"extranonce1_size": 8, "extranonce2_size": 8, )"
                                        R"("coinbase_signature": ")") +
                            std::string(74, 'x') + R"("})"));
}

TEST_CASE("rpc_endpoints lists the primary first, then failover in order") {
    const auto endpoints = Config::from_string(R"({
        "bitcoin_nodes": [
            {"address": "p", "username": "pu", "password": "pp"},
            {"address": "a", "username": "au", "password": "ap"},
            {"address": "b", "username": "bu", "password": "bp"}
        ]
    })").rpc_endpoints();
    REQUIRE(endpoints.size() == 3);
    CHECK(endpoints[0].url == "p");
    CHECK(endpoints[0].password == "pp");
    CHECK(endpoints[1].url == "a");
    CHECK(endpoints[2].url == "b");
}

File: python/erikslund_pool/tests/functional/test_config.py
"""Functional tests for Settings config parsing (conf/pool.schema.json).

Config files load as YAML (a superset of JSON); fixtures are written as YAML.
"""
from __future__ import annotations

import os
import tempfile

import yaml

from erikslund_pool.config import Settings
from erikslund_pool.exceptions import ConfigError
from erikslund_pool.tests.base import SoloPoolTestCase


class TestSettings(SoloPoolTestCase):
    def _write(self, obj) -> str:
        fd, path = tempfile.mkstemp(suffix=".yaml")
        with os.fdopen(fd, "w", encoding="utf-8") as f:
            yaml.safe_dump(obj, f)
        self.addCleanup(os.unlink, path)
        return path

    def test_load_overrides_and_keeps_defaults(self):
        path = self._write({"stratum_listen": ["0.0.0.0:4444"], "initial_difficulty": 8})
        config = Settings.load(path)
        self.assertEqual(config.bind_port, 4444)
        self.assertEqual(config.initial_difficulty, 8)
        self.assertEqual(config.rpc_user, "erikslund")  # untouched default

    def test_unknown_key_rejected(self):
        path = self._write({"bogus_key": 1})
        with self.assertRaises(ConfigError):
            Settings.load(path)

    def test_missing_file_raises_config_error(self):
        with self.assertRaises(ConfigError):
            Settings.load("/nonexistent/erikslund_pool.json")

    def test_non_object_json_raises_config_error(self):
        path = self._write([1, 2, 3])
        with self.assertRaises(ConfigError):
            Settings.load(path)


class TestConfigSchema(SoloPoolTestCase):
    def test_schema_parses(self):
        config = Settings.from_dict({
            "$schema": "../conf/pool.schema.json",
            "bitcoin_nodes": [
                {"address": "bitcoind:8332", "username": "u", "password": "p", "notify": True},
                {"address": "backup:8332", "username": "u2", "password": "p2"},
            ],
            "stratum_listen": ["0.0.0.0:3333", "0.0.0.0:4001"],
            "coinbase_signature": "/erikslund/",
            "initial_difficulty": 10000,
            "minimum_difficulty": 1,
            "maximum_difficulty": 1000000,
            "extranonce1_size": 4,
            "extranonce2_size": 8,
            "api_listen": ["127.0.0.1:7777"],
            "zmq_block_endpoint": "tcp://bitcoind:28332",
            "block_poll_milliseconds": 250,
            "work_rebroadcast_seconds": 20,
            "max_clients": 50,
            "drop_idle_seconds": 900,
            "version_rolling_mask": "1fffe000",
        })
        self.assertEqual(config.rpc_url, "bitcoind:8332")
        self.assertEqual(config.rpc_user, "u")
        self.assertEqual(config.rpc_password, "p")
        self.assertEqual(len(config.rpc_failover), 1)
        self.assertEqual(config.rpc_failover[0]["url"], "backup:8332")
        self.assertEqual(config.bind_host, "0.0.0.0")
        self.assertEqual(config.bind_ports, [3333, 4001])
        self.assertEqual(config.bind_port, 3333)
        self.assertEqual(config.coinbase_signature, "/erikslund/")
        self.assertEqual(config.initial_difficulty, 10000)
        self.assertEqual(config.minimum_difficulty, 1)
        self.assertEqual(config.maximum_difficulty, 1000000)
        self.assertEqual(config.extranonce1_size, 4)
        self.assertEqual(config.extranonce2_size, 8)
        self.assertEqual(config.api_host, "127.0.0.1")
        self.assertEqual(config.api_port, 7777)
        self.assertEqual(config.zmq_block_endpoint, "tcp://bitcoind:28332")
        self.assertAlmostEqual(config.poll_interval, 0.25)   # block_poll_milliseconds -> seconds
        self.assertEqual(config.work_rebroadcast_seconds, 20)
        self.assertEqual(config.max_clients, 50)
        self.assertEqual(config.drop_idle_seconds, 900)
        self.assertEqual(config.version_rolling_mask, 0x1FFFE000)

    def test_single_stratum_listen_string(self):
        config = Settings.from_dict({"stratum_listen": "0.0.0.0:3333"})
        self.assertEqual(config.bind_host, "0.0.0.0")
        self.assertEqual(config.bind_ports, [3333])
        self.assertEqual(config.bind_port, 3333)

    def test_flat_schema_is_rejected(self):
        # A flat (non-nested) schema is not accepted.
        with self.assertRaises(ConfigError):
            Settings.from_dict({"rpc_url": "http://node:8332", "bind_port": 4444})

    def test_pool_extra_knobs_pass_through(self):
        config = Settings.from_dict({"vardiff_target_shares_per_minute": 30, "status_interval_seconds": 45.0})
        self.assertEqual(config.vardiff_target_shares_per_minute, 30)
        self.assertEqual(config.status_interval_seconds, 45.0)

    def test_empty_dict_keeps_all_defaults(self):
        config = Settings.from_dict({})
        defaults = Settings()
        self.assertEqual(config.rpc_url, defaults.rpc_url)
        self.assertEqual(config.bind_port, defaults.bind_port)
        self.assertEqual(config.api_port, defaults.api_port)
        self.assertEqual(config.version_rolling_mask, defaults.version_rolling_mask)

    def test_three_node_failover_ordering(self):
        # nodes[0] is primary; the rest are failover in declared order; missing
        # credentials yield empty user/password.
        config = Settings.from_dict({"bitcoin_nodes": [
            {"address": "a:1", "username": "ua", "password": "pa"},
            {"address": "b:2"},
            {"address": "c:3", "username": "uc", "password": "pc"},
        ]})
        self.assertEqual(config.rpc_url, "a:1")
        self.assertEqual(config.rpc_user, "ua")
        self.assertEqual([f["url"] for f in config.rpc_failover], ["b:2", "c:3"])
        self.assertEqual(config.rpc_failover[0], {"url": "b:2", "user": "", "password": ""})
        self.assertEqual(config.rpc_failover[1]["user"], "uc")

    def test_single_node_has_no_failover(self):
        config = Settings.from_dict({"bitcoin_nodes": [{"address": "only:8332"}]})
        self.assertEqual(config.rpc_url, "only:8332")
        self.assertEqual(config.rpc_failover, [])

    def test_api_listen_bare_string(self):
        config = Settings.from_dict({"api_listen": "0.0.0.0:9999"})
        self.assertEqual(config.api_host, "0.0.0.0")
        self.assertEqual(config.api_port, 9999)

    def test_stratum_listen_host_defaults_when_omitted(self):
        # ":3333" has no host component -> falls back to 0.0.0.0.
        config = Settings.from_dict({"stratum_listen": ":3333"})
        self.assertEqual(config.bind_host, "0.0.0.0")
        self.assertEqual(config.bind_ports, [3333])

    def test_version_rolling_mask_accepts_int(self):
        config = Settings.from_dict({"version_rolling_mask": 0x00FF0000})
        self.assertEqual(config.version_rolling_mask, 0x00FF0000)

    def test_max_workers_per_address_loads_and_rejects_negative(self):
        self.assertEqual(
            Settings.from_dict({"max_workers_per_address": 32}).max_workers_per_address, 32)
        self.assertEqual(Settings.from_dict({}).max_workers_per_address, 256)  # default
        self.assertEqual(
            Settings.from_dict({"max_workers_per_address": 0}).max_workers_per_address, 0)
        with self.assertRaises(ConfigError):
            Settings.from_dict({"max_workers_per_address": -1})

    def test_user_stats_retention_days_loads_and_rejects_negative(self):
        self.assertEqual(
            Settings.from_dict({"user_stats_retention_days": 30}).user_stats_retention_days, 30)
        self.assertEqual(Settings.from_dict({}).user_stats_retention_days, 90)  # default
        self.assertEqual(
            Settings.from_dict({"user_stats_retention_days": 0}).user_stats_retention_days, 0)
        with self.assertRaises(ConfigError):
            Settings.from_dict({"user_stats_retention_days": -1})
        # Upper bound (~100 years): beyond it the cutoff arithmetic would overflow.
        with self.assertRaises(ConfigError):
            Settings.from_dict({"user_stats_retention_days": 36501})

    def test_version_rolling_mask_is_clamped_to_the_bip320_range(self):
        self.assertEqual(
            Settings.from_dict({"version_rolling_mask": 0xFFFFFFFF}).version_rolling_mask,
            0x1FFFE000)
        self.assertEqual(
            Settings.from_dict({"version_rolling_mask": 0x00003000}).version_rolling_mask,
            0x00002000)

    def test_block_poll_milliseconds_converts_to_seconds(self):
        self.assertAlmostEqual(
            Settings.from_dict({"block_poll_milliseconds": 1500}).poll_interval, 1.5)
        self.assertAlmostEqual(
            Settings.from_dict({"block_poll_milliseconds": 250}).poll_interval, 0.25)
        # 0 ms is rejected: it would busy-loop the poller.
        with self.assertRaises(ConfigError):
            Settings.from_dict({"block_poll_milliseconds": 0})

    def test_donation_fields_and_boolean_flags(self):
        config = Settings.from_dict({
            "donation_percent": 1.5, "donation_address": "bc1qexample",
            "fast_block_notify": False, "variable_difficulty": False,
        })
        self.assertEqual(config.donation_percent, 1.5)
        self.assertEqual(config.donation_address, "bc1qexample")
        self.assertFalse(config.fast_block_notify)
        self.assertFalse(config.variable_difficulty)

    def test_worker_threads_and_size_knobs(self):
        config = Settings.from_dict({
            "worker_threads": 8, "max_line_bytes": 65536, "drop_idle_seconds": 0,
            "auth_timeout_seconds": 45, "max_protocol_errors": 7,
            "extranonce1_size": 6, "extranonce2_size": 4, "coinbase_version": 2,
        })
        self.assertEqual(config.worker_threads, 8)
        self.assertEqual(config.max_line_bytes, 65536)
        self.assertEqual(config.drop_idle_seconds, 0)
        self.assertEqual(config.auth_timeout_seconds, 45)
        self.assertEqual(config.max_protocol_errors, 7)
        self.assertEqual(config.extranonce1_size, 6)
        self.assertEqual(config.extranonce2_size, 4)
        self.assertEqual(config.coinbase_version, 2)

    def test_schema_hint_key_is_ignored(self):
        # The "$schema" editor hint produces no field.
        config = Settings.from_dict({"$schema": "../conf/pool.schema.json"})
        self.assertEqual(config.bind_port, Settings().bind_port)

    def test_unknown_key_names_are_reported(self):
        with self.assertRaises(ConfigError) as ctx:
            Settings.from_dict({"totally_made_up": 1, "another_bad": 2})
        message = str(ctx.exception)
        self.assertIn("another_bad", message)
        self.assertIn("totally_made_up", message)

    def test_non_mapping_rejected(self):
        for bad in ([1, 2, 3], "a string", 42):
            with self.assertRaises(ConfigError):
                Settings.from_dict(bad)


class TestConfigValidation(SoloPoolTestCase):
    """Schema range bounds are enforced.

    These values would otherwise load and then busy-loop, divide by zero, or
    silently reject every share.
    """

    def test_out_of_range_values_rejected(self):
        for cfg in (
            {"initial_difficulty": 0}, {"initial_difficulty": -1},
            {"minimum_difficulty": 0}, {"minimum_difficulty": -0.5},
            {"maximum_difficulty": -0.1},
            {"vardiff_target_shares_per_minute": 0},
            {"vardiff_retarget_seconds": 0},
            {"work_rebroadcast_seconds": 0},
            {"block_poll_milliseconds": 0},
            # extranonce1 min is 4: a 2-3 byte space can wrap and hand two miners the
            # same search space.
            {"extranonce1_size": 1}, {"extranonce1_size": 2},
            {"extranonce1_size": 3}, {"extranonce1_size": 9},
            {"extranonce2_size": 0}, {"extranonce2_size": 9},
            {"donation_percent": -1}, {"donation_percent": 101},
            {"max_clients": -1}, {"drop_idle_seconds": -1},
            {"auth_timeout_seconds": -1}, {"max_protocol_errors": -1},
            {"status_interval_seconds": -1},
        ):
            with self.subTest(cfg=cfg), self.assertRaises(ConfigError):
                Settings.from_dict(cfg)

    def test_boundary_values_accepted(self):
        for cfg in (
            {"extranonce1_size": 4}, {"extranonce1_size": 8}, {"extranonce2_size": 8},
            {"vardiff_retarget_seconds": 1}, {"work_rebroadcast_seconds": 1},
            {"block_poll_milliseconds": 1}, {"maximum_difficulty": 0},
            {"drop_idle_seconds": 0}, {"auth_timeout_seconds": 0},
            {"status_interval_seconds": 0}, {"donation_percent": 0},
            {"donation_percent": 100},
        ):
            with self.subTest(cfg=cfg):
                Settings.from_dict(cfg)  # must not raise
